Change some comments into proper docstrings

DilumAluthge opened the pull request:
This PR takes the comments associated with some generic function stubs and turns them into proper docstrings that will show up correctly in documentation builds, etc.

cc: @tlienart 
cc: @ablaom

DilumAluthge committed:
Change some comments into proper docstrings

tlienart commented:
Thanks @DilumAluthge! 🙏  

I'm a bit ambivalent so will leave the decision to Anthony, the thing is that we define these stubs here, then implement them effectively in MLJBase where the docstring should be (and, in fact, users should have access to  that docstring). No  one is expected to just load `MLJModelInterface` and query its functions, if you're looking at that repo, it means you're writing an interface and in that case you should look at the code.

Anyway, the origin of your PR may be that the docstring of `fit`, `update` `predict_mean` etc do not work properly, indeed if you load `MLJBase` and  do `?fit` you won't get anything useful (that's not good).

In order for those to work I imagine we might be forced to have the detailed docstrings in MLJModelInterface as opposed to elsewhere?

(apologies for the rambling)

DilumAluthge commented:
> Anyway, the origin of your PR may be that the docstring of `fit`, `update` `predict_mean` etc do not work properly, indeed if you load `MLJBase` and do `?fit` you won't get anything useful (that's not good).

Yep this is one of the core motivations for this PR.

The `fit` function is owned by `MLJModelInterface`, so in order for `?fit` to work correctly, you need to have a docstring for `MLJModelInterface.fit`. And the docstring needs to be in the `MLJModelInterface` repo.

ablaom commented:
Okay, thanks @DilumAluthge . This is useful.

With the exception of `fitted_params` and `save`, this PR only adds docstrings to functions that don't already have docstrings further up the MLJ food chain. Moreover, I don't think the MLJ *user* needs doc strings for these latter methods, beyond what is provided here. 

In the case of the two the two exceptions, the existing doc-strings refer to different public *methods* (signatures). However, to avoid clutter for the general user, I would recommend we don't add these strings for the "private" methods.

@DilumAluthge Would you consider dropping just those two?

DilumAluthge committed:
Remove docstrings for `fitted_params`, `save`, and `restore`

DilumAluthge commented:
> @DilumAluthge Would you consider dropping just those two?

Done! I removed the docstrings for `fitted_params` and `save`.

It seems like `save` and `restore` go together. So I also removed the docstring for `restore`. But I can put it back if you want - just let me know.

ablaom commented:
> But I can put it back if you want - just let me know.

No, no. I just missed that one. Good catch.

And thanks again.

DilumAluthge commented:
Thanks @ablaom and @tlienart!

Any chance we could get a new release?

ablaom commented:
@DilumAluthge  Oops. This should have been based onto dev, not master. Please remember for next time 😄 

https://github.com/alan-turing-institute/MLJ.jl/blob/master/CONTRIBUTING.md

DilumAluthge commented:
> @DilumAluthge Oops. This should have been based onto dev, not master. Please remember for next time 😄
> 
> https://github.com/alan-turing-institute/MLJ.jl/blob/master/CONTRIBUTING.md

Whoops, my mistake!

FYI, if you want, you can go to the GitHub settings for the repo and set the "Default Branch" for the repo to be the `dev` branch. The advantage of making the `dev` branch the default branch is that GitHub will automatically open pull requests against `dev` by default.

Of course the correct answer is that contributors like myself should read the CONTRIBUTING.md file 😄
